p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java

```diff
@@ -0,0 +1,190 @@
+
+            history = new HashMap<>();
+            store = readStore(getFileName());
+            resourceFiles.forEach(file -> {
+                SplitStore tmp = readStore(file.getName().replace(postFix, ""));
+                history.put(file.getName().replace(postFix, "").replace(getFileName(), "").replace("_", ""), tmp);
+            });
+        }
+    }
+
+    private T readStore(String name) {
+        T store = storage.initAndGetPersistedWithFileName(name, 100);
+        if (store != null) {
+            log.info("{}: size of {}: {} MB", this.getClass().getSimpleName(),
+                    storage.getClass().getSimpleName(),
+                    store.toProtoMessage().toByteArray().length / 1_000_000D);
+        } else {
+            store = createStore();
+        }
+
+        return store;
+    }
+
+    @Override
+    protected void makeFileFromResourceFile(String postFix) {
+        File dbDir = new File(absolutePathOfStorageDir);
+        if (!dbDir.exists() && !dbDir.mkdir())
+            log.warn("make dir failed.\ndbDir=" + dbDir.getAbsolutePath());
+
+        // check resources for files
+        List<String> versions = new ArrayList<>();
+        versions.add(Version.VERSION);
+        versions.addAll(Version.history);
+        List<String> resourceFiles = versions.stream().map(s -> getFileName() + "_" + s + postFix).collect(Collectors.toList());
+
+        // if not, copy and split
+        resourceFiles.forEach(file -> {
+            final File destinationFile = new File(Paths.get(absolutePathOfStorageDir, file.replace(postFix, "")).toString());
+            final String resourceFileName = file;
+            if (!destinationFile.exists()) {
+                try {
+                    log.info("We copy resource to file: resourceFileName={}, destinationFile={}", resourceFileName, destinationFile);
+                    FileUtil.resourceToFile(resourceFileName, destinationFile);
+                } catch (ResourceNotFoundException e) {
+                    log.info("Could not find resourceFile " + resourceFileName + ". That is expected if none is provided yet.");
+                } catch (Throwable e) {
+                    log.error("Could not copy resourceFile " + resourceFileName + " to " +
+                            destinationFile.getAbsolutePath() + ".\n" + e.getMessage());
+                    e.printStackTrace();
+                }
+            } else {
+                log.debug(file + " file exists already.");
+            }
+        });
+
+        // split
+        // - get all
+        history = new HashMap<>();
+        store = readStore(getFileName());
+        resourceFiles.forEach(file -> {
+            SplitStore tmp = readStore(file.replace(postFix, ""));
+            history.put(file.replace(postFix, "").replace(getFileName(), "").replace("_", ""), tmp);
+            // - subtract all that is in resource files
+            store.getMap().keySet().removeAll(tmp.getMap().keySet());
```

Comment:
With this change I think we're making the resource files the final arbiter of what is the distributed data. Is that what we want? Even for full nodes that are always on.

From what I understood so far, nodes will keep getting data from the network. With new versions all new data since the last version will be stored in a new `resource_x.x.x_MAINNET` file and all nodes will restore this data from resources. It looks like it will just throw away its own local data on new versions though, that seems incorrect.

Reply:
> With this change I think we're making the resource files the final arbiter of what is the distributed data. Is that what we want? Even for full nodes that are always on.

This seems to be the only way of introducing "checkpoints" so that we do not have to query all the data.

> It looks like it will just throw away its own local data on new versions though, that seems incorrect.

It does not throw away any data. It only de-duplicates the data. If we have a data store "incoming" from resources, we remove the data that is contained in the resource store from our live data store.

An example:
Given _ResourceStore={1,3}_ and _LiveStore(t)={1,2,3,4,5}_ and we enter the copy and split business logic, the outcome would be having the _ResourceStore_ being copied to the working directory and the new _LiveStore(t+1)=LiveStore(t) \ ResourceStore = {2,4,5}_. Together, they hold {1,2,3,4,5}.

Bringing it all together:
In terms of initial data request size, we can now ask for all data except _ResourceStore_ simply by telling other that we have the "ResourceStore_1.3.4" -> sending {"1.3.4"} instead of {1,3}. And because in real live, the data stores count >100k objects, we save a whole lot of data. Additionally, we actually ask for all data "since" (instead of "except") and thus keep the request object count at O(1) in the future.